File: src/main/java/org/modelingvalue/nelumbo/patterns/SequencePattern.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.patterns;

import java.io.Serial;
import java.util.function.Function;

import org.modelingvalue.collections.List;
import org.modelingvalue.collections.Map;
import org.modelingvalue.collections.mutable.MutableList;
import org.modelingvalue.nelumbo.AstElement;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.Type;
import org.modelingvalue.nelumbo.Variable;
import org.modelingvalue.nelumbo.syntax.ParseState;
import org.modelingvalue.nelumbo.syntax.Token;
import org.modelingvalue.nelumbo.syntax.TokenType;

public class SequencePattern extends Pattern {
    @Serial
    private static final long serialVersionUID = 1477171023667359130L;

    public SequencePattern(Type type, List<AstElement> elements, List<Pattern> patterns) {
        super(type, elements, patterns);
    }

    protected SequencePattern(Object[] args, SequencePattern declaration) {
        super(args, declaration);
    }

    @Override
    protected SequencePattern struct(Object[] array, Node declaration) {
        return new SequencePattern(array, (SequencePattern) declaration);
    }

    @SuppressWarnings("unchecked")
    public List<Pattern> elements() {
        return (List<Pattern>) get(0);
    }

    @Override
    public String name() {
        StringBuilder name = new StringBuilder();
        for (Pattern element : elements()) {
            name.append(element.name());
        }
        return name.toString();
    }

    @Override
    public String toString(TokenType[] previous) {
        return elements().map(Object::toString).reduce("", (a, b) -> a + b);
    }

    @Override
    public Pattern setPresedence(int precedence) {
        List<Pattern> elements = elements();
        for (int i = 0; i < elements.size(); i++) {
            Pattern pa = elements.get(i);
            Pattern pb = pa.setPresedence(precedence);
            if (!pb.equals(pa)) {
                elements = elements.replace(i, pb);
            }
        }
        return set(0, elements);
    }

    @Override
    public Pattern setTypes(Function<Type, Type> typeFunction) {
        List<Pattern> elements = elements();
        for (int i = 0; i < elements.size(); i++) {
            Pattern pa = elements.get(i);
            Pattern pb = pa.setTypes(typeFunction);
            if (!pb.equals(pa)) {
                elements = elements.replace(i, pb);
            }
        }
        return set(0, elements);
    }

    @Override
    public ParseState state(ParseState state) {
        for (Pattern element : elements().reverse()) {
            state = element.state(state);
        }
        return state;
    }

    @Override
    public List<Type> argTypes(List<Type> types) {
        for (Pattern element : elements()) {
            types = element.argTypes(types);
        }
        return types;
    }

    @SuppressWarnings("unchecked")
    @Override
    protected int string(List<Object> args, int ai, StringBuffer sb, TokenType[] previous, boolean alt) {
        if (ai < 0 || args.size() <= ai) {
            return -1;
        }
        if (argTypes(List.of()).size() == 1) {
            args = List.of(List.of(args.get(ai)));
        }
        if (args.get(ai) instanceof List list) {
            StringBuffer inner = new StringBuffer();
            int ii = 0;
            for (Pattern element : elements()) {
                ii = element.string(list, ii, inner, previous, false);
                if (ii < 0) {
                    return -1;
                }
            }
            sb.append(inner);
            return ai + 1;
        }
        return -1;
    }

    @Override
    protected int args(List<AstElement> elements, int i, MutableList<Object> args, boolean alt, Functor functor, Map<Variable, Type> typeArgs) {
        List<Object> result = List.of();
        for (Pattern element : elements()) {
            MutableList<Object> inner = MutableList.of(List.of());
            int ii = element.args(elements, i, inner, false, functor, typeArgs);
            if (ii >= 0) {
                result = result.addAll(inner.toImmutable());
                i = ii;
            } else {
                return -1;
            }
        }
        args.add(result.size() > 1 ? result : result.first());
        return i;
    }

    @Override
    public Pattern declaration(Token token) {
        for (Pattern element : elements()) {
            Pattern decl = element.declaration(token);
            if (decl != null) {
                return decl;
            }
        }
        return null;
    }

}

File: web-ts/src/patterns/SequencePattern.ts
/**
 * SequencePattern - matches a sequence of patterns.
 * Ported from Java: org.modelingvalue.nelumbo.patterns.SequencePattern
 */

import { List, Map } from 'immutable';
import { TokenType } from '../TokenType';
import type { Token } from '../Token';
import type { AstElement } from '../core/AstElement';
import { Type } from '../core/Type';
import { Variable } from '../core/Variable';
import { Node } from '../core/Node';
import { Pattern } from './Pattern';
import type { ParseState } from '../syntax/ParseState';
import type { Functor } from './Functor';

export class SequencePattern extends Pattern {
  constructor(type: Type, elements: List<AstElement>, ...args: unknown[]) {
    super(type, elements, ...args);
  }

  protected static fromData(data: unknown[], declaration?: Node): SequencePattern {
    const pattern = Object.create(SequencePattern.prototype) as SequencePattern;
    (pattern as unknown as { _data: unknown[] })._data = data;
    (pattern as unknown as { _declaration: Node })._declaration = declaration ?? pattern;
    return pattern;
  }

  protected struct(data: unknown[], declaration?: Node): SequencePattern {
    return SequencePattern.fromData(data, declaration ?? this.declaration());
  }

  patternElements(): List<Pattern> {
    return this.get(0) as List<Pattern>;
  }

  name(): string {
    let name = '';
    for (const element of this.patternElements()) {
      name += element.name();
    }
    return name;
  }

  toString(_previous?: TokenType[]): string {
    return this.patternElements().map(e => String(e)).join('');
  }

  setPrecedence(precedence: number): Pattern {
    let elems = this.patternElements();
    for (let i = 0; i < elems.size; i++) {
      const pa = elems.get(i)!;
      const pb = pa.setPrecedence(precedence);
      if (!pb.equals(pa)) {
        elems = elems.set(i, pb);
      }
    }
    return this.set(0, elems) as SequencePattern;
  }

  setTypes(typeFunction: (type: Type) => Type): Pattern {
    let elems = this.patternElements();
    for (let i = 0; i < elems.size; i++) {
      const pa = elems.get(i)!;
      const pb = pa.setTypes(typeFunction);
      if (!pb.equals(pa)) {
        elems = elems.set(i, pb);
      }
    }
    return this.set(0, elems) as SequencePattern;
  }

  parseState(next: ParseState, functor: Functor): ParseState {
    let state = next;
    for (const element of this.patternElements().reverse()) {
      state = element.parseState(state, functor);
    }
    return state;
  }

  argTypes(types: List<Type>): List<Type> {
    for (const element of this.patternElements()) {
      types = element.argTypes(types);
    }
    return types;
  }

  string(args: List<unknown>, ai: number, sb: string[], previous: TokenType[], _alt: boolean): number {
    let argList = args;
    if (this.argTypes(List()).size === 1) {
      argList = List([List([args.get(ai)])]);
    }

    const listArg = argList.get(ai);
    if (List.isList(listArg)) {
      const list = listArg as List<unknown>;
      const inner: string[] = [];
      let ii = 0;
      for (const element of this.patternElements()) {
        ii = element.string(list, ii, inner, previous, false);
        if (ii < 0) {
          return -1;
        }
      }
      sb.push(...inner);
      return ai + 1;
    }
    return -1;
  }

  extractArgs(
    astElements: List<AstElement>,
    i: number,
    args: unknown[],
    _alt: boolean,
    functor: Functor,
    typeArgs: Map<Variable, Type>
  ): number {
    const result: unknown[] = [];
    for (const element of this.patternElements()) {
      const inner: unknown[] = [];
      const ii = element.extractArgs(astElements, i, inner, false, functor, typeArgs);
      if (ii >= 0) {
        result.push(...inner);
        i = ii;
      } else {
        return -1;
      }
    }
    args.push(result.length > 1 ? List(result) : result[0]);
    return i;
  }

  tokenDeclaration(token: Token): Pattern | null {
    for (const element of this.patternElements()) {
      const decl = element.tokenDeclaration(token);
      if (decl !== null) {
        return decl;
      }
    }
    return null;
  }
}
